# Goals

### 1. Understand what single label image classifications is
        - One label per image

### 2. Understand single label - csv type data structure

## Table of Contents


## [0. Install](#0)


## [1. Load train data - csv type](#1)


## [2. Understand experiment data-summary](#2)


## [3. Load val data along with train data](#3)


## [4. Understand experiment data-summary](#4)

<a id='0'></a>
# Install Monk
  
 - git clone https://github.com/Tessellate-Imaging/monk_v1.git
 
 - cd monk_v1/installation && pip install -r requirements_cu9.txt
     - (Select the requirements file as per OS and CUDA version)

In [3]:
!git clone https://github.com/Tessellate-Imaging/monk_v1.git

Cloning into 'monk_v1'...
remote: Enumerating objects: 11, done.
remote: Counting objects: 100% (11/11), done.
remote: Compressing objects: 100% (7/7), done.
remote: Total 1916 (delta 4), reused 10 (delta 4), pack-reused 1905
Receiving objects: 100% (1916/1916), 69.95 MiB | 10.27 MiB/s, done.
Resolving deltas: 100% (1024/1024), done.


In [ ]:
# Select the requirements file as per OS and CUDA version
!cd monk_v1/installation && pip install -r requirements_cu9.txt

# Imports

In [1]:
# Monk
import os
import sys
sys.path.append("monk_v1/monk/");

In [2]:
#Using mxnet-gluon backend 
from gluon_prototype import prototype

## Creating and managing experiments
    - Provide project name
    - Provide experiment name

In [5]:
gtf = prototype(verbose=1);
gtf.Prototype("Dataset-Type", "exp-csv");

Mxnet Version: 1.5.1

Experiment Details
    Project: Dataset-Type
    Experiment: exp-csv
    Dir: /home/abhi/Desktop/Work/tess_tool/gui/v0.3/finetune_models/Organization/development/v5.0_blocks/study_roadmap/acceptable_datatypes_in_monk/workspace/Dataset-Type/exp-csv/



### This creates files and directories as per the following structure
    
    
    workspace
        |
        |--------Dataset-Type 
                        |
                        |
                        |-----exp-foldered 
                                    |
                                    |-----experiment-state.json
                                    |
                                    |-----output
                                            |
                                            |------logs (All training logs and graphs saved here)
                                            |
                                            |------models (all trained models saved here)

<a id='1'></a>
# Load train data - foldered type

## Quick mode training

    - Using Default Function
        - dataset_path
        - model_name
        - num_epochs
        
        
## Dataset folder structure

    parent_directory
        |
        |
        |------train (folder can be named anything)
                |------img1.jpg
                |------img2.jpg
                |------.... (and so on)
        |
        |------train_labels.csv (file can be named anything)
        
        
    train_labels.csv has 2 columns (column names could be anything)
        - ID                 | Labels
          img1.jpg           |  cat
          img2.jpg           |  dog
          img3.jpg           |  cat
          img4.jpg           |  dog
                ..... (and so on)
                
### In similar fashion you can structure your own dataset

### Along with providing dataset_path you now provide path_to_csv

In [6]:

gtf.Default(dataset_path="monk_v1/monk/system_check_tests/datasets/dataset_csv_id/train",
            path_to_csv="monk_v1/monk/system_check_tests/datasets/dataset_csv_id/train.csv",
            model_name="resnet18_v1", 
            num_epochs=5);

Dataset Details
    Train path:     monk_v1/monk/system_check_tests/datasets/dataset_csv_id/train
    Val path:       None
    CSV train path: monk_v1/monk/system_check_tests/datasets/dataset_csv_id/train.csv
    CSV val path:   None

Dataset Params
    Input Size:   224
    Batch Size:   4
    Data Shuffle: True
    Processors:   4
    Train-val split:   0.7
    Delimiter:   ,

Pre-Composed Train Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Pre-Composed Val Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Dataset Numbers
    Num train images: 70
    Num val images:   30
    Num classes:      2

Model Params
    Model name:           resnet18_v1
    Use Gpu:              True
    Use pretrained:       True
    Freeze base network:  True

Model Details
    Loading pretrained model
    Model Loaded on device
        Model name

monk_v1/monk/system/imports.py:193: UserWarning: ArgumentWarning: clipnorm and clipvalue are active only for keras in current version of Monk
  warnings.warn(msg)
monk_v1/monk/system/imports.py:193: UserWarning: ArgumentWarning: momentum_dampening_rate is active only for pytorch in current version of Monk
  warnings.warn(msg)


<a id='2'></a>
# Understand experiment data-summary

    Dataset Details
        Train path:     monk_v1/monk/system_check_tests/datasets/dataset_csv_id/train
        Val path:       None
        CSV train path: monk_v1/monk/system_check_tests/datasets/dataset_csv_id/train.csv
        CSV val path:   None
        
        
### Train path is mentioned and Val path is None

### CSV Train path is mentioned unlike in foldered type dataset


    Dataset Parameters:
        Input Size:   224
        Batch Size:   4
        Shuffle:      True
        Processors:   4
        Num Classes:  2

###  Data input size is 224 (default setting)
###  Batch size is 4 (default setting)

    Dataset Numbers
        Num train images: 70
        Num val images:   30
        Num classes:      2
        
### Since val data is missing, train data is split into train and val with a ratio 70:30 rule
    - 70% of data being used for training
    - 30% of data being used for validation
    
### Auto detect number of classes

<a id='3'></a>
# Load val data along with train data - foldered type

## Quick mode training

    - Using Default Function
        - dataset_path
        - model_name
        - num_epochs
        
        
## Dataset folder structure

    parent_directory
        |
        |
        |------train (folder can be named anything)
                |------img1.jpg
                |------img2.jpg
                |------.... (and so on)
        |
        |------train_labels.csv (file can be named anything)
        
        
    train_labels.csv has 2 columns (column names could be anything)
        - ID                 | Labels
          img1.jpg           |  cat
          img2.jpg           |  dog
          img3.jpg           |  cat
          img4.jpg           |  dog
                ..... (and so on)
                
                
    parent_directory
        |
        |
        |------val (folder can be named anything)
                |------img1.jpg
                |------img2.jpg
                |------.... (and so on)
        |
        |------val_labels.csv (file can be named anything)
        
        
    val_labels.csv has 2 columns (column names could be anything)
        - ID                 | Labels
          img1.jpg           |  cat
          img2.jpg           |  dog
          img3.jpg           |  cat
          img4.jpg           |  dog
                ..... (and so on)
                
                
### In similar fashion you can structure your own dataset

### Instead of providing dataset as string, provide list of strings
        - First string: Train data path
        - Second string: Validation data path
        
        
### Alongside the same goes with path_to_csv
        - First string: Train data csv path
        - Second string: Validation data csv path

In [7]:
gtf.Default(dataset_path=["monk_v1/monk/system_check_tests/datasets/dataset_csv_id/train",
                          "monk_v1/monk/system_check_tests/datasets/dataset_csv_id/val"],
            path_to_csv=["monk_v1/monk/system_check_tests/datasets/dataset_csv_id/train.csv",
                         "monk_v1/monk/system_check_tests/datasets/dataset_csv_id/val.csv"],
            model_name="resnet18_v1", 
            num_epochs=5);

Dataset Details
    Train path:     monk_v1/monk/system_check_tests/datasets/dataset_csv_id/train
    Val path:       monk_v1/monk/system_check_tests/datasets/dataset_csv_id/val
    CSV train path: monk_v1/monk/system_check_tests/datasets/dataset_csv_id/train.csv
    CSV val path:   monk_v1/monk/system_check_tests/datasets/dataset_csv_id/val.csv

Dataset Params
    Input Size:   224
    Batch Size:   4
    Data Shuffle: True
    Processors:   4
    Delimiter:   ,

Pre-Composed Train Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}, {'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Pre-Composed Val Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}, {'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Datas

monk_v1/monk/system/imports.py:193: UserWarning: ArgumentWarning: clipnorm and clipvalue are active only for keras in current version of Monk
  warnings.warn(msg)
monk_v1/monk/system/imports.py:193: UserWarning: ArgumentWarning: momentum_dampening_rate is active only for pytorch in current version of Monk
  warnings.warn(msg)


<a id='4'></a>
# Understand experiment data-summary

    Dataset Details
        Train path:     monk_v1/monk/system_check_tests/datasets/dataset_csv_id/train
        Val path:       monk_v1/monk/system_check_tests/datasets/dataset_csv_id/val
        CSV train path: monk_v1/monk/system_check_tests/datasets/dataset_csv_id/train.csv
        CSV val path:   monk_v1/monk/system_check_tests/datasets/dataset_csv_id/val.csv
        
        
### Both Train and Val path are mentioned

### Both CSV Train and val path are mentioned unlike in foldered type dataset

    Dataset Parameters:
        Input Size:   224
        Batch Size:   4
        Shuffle:      True
        Processors:   4
        Num Classes:  2

###  Data input size is 224 (default setting)
###  Batch size is 4 (default setting)

    Dataset Numbers
        Num train images: 100
        Num val images:   50
        Num classes:      2
        
### Since val data is present, train data is not split further
    - all 100 images of data being used for training
    - all 50 images of data being used for validation
    
### Auto detect number of classes